drivers/samsung/displays/mdc_protocol.cr

```diff
@@ -0,0 +1,376 @@
+    if val
+      if !self[:audio_mute]?
+        self[:audio_mute] = true
+        @previous_volume = self[:volume].as_i
+        volume(0)
+      end
+    else
+      unmute_audio
+    end
+  end
+
+  def unmute_audio
+    if self[:audio_mute]?
+      self[:audio_mute] = false
+      volume(@previous_volume)
+    end
+  end
+
+  # check software version
+  def software_version?
+    do_send(Command::Software_Version)
+  end
+
+  def serial_number?
+    do_send(Command::Serial_Number)
+  end
+
+  def switch_to(input : Input, **options)
+    @input_stable = false
+    @input_target = input
+    do_send(Command::Input, input.value, **options)
+  end
+
+  enum SpeakerModes
+    Internal = 0
+    External = 1
+  end
+
+  def speaker_select(mode : String, **options)
+    do_send(Command::Speaker, SpeakerModes.parse(mode).value, **options)
+  end
+
+  def do_poll
+    do_send(Command::Status, Bytes.empty, priority: 0)
+    power? unless self[:hard_off]?
+  end
+
+  DEVICE_SETTINGS = {
+    network_standby: Bool,
+    auto_off_timer: Bool,
+    auto_power: Bool,
+    volume: Int32,
+    contrast: Int32,
+    brightness: Int32,
+    sharpness: Int32,
+    colour: Int32,
+    tint: Int32,
+    red_gain: Int32,
+    green_gain: Int32,
+    blue_gain: Int32
+  }
+  {% for name, kind in DEVICE_SETTINGS %}
+    @[Security(Level::Administrator)]
+    def {{name.id}}(value : {{kind}}, **options)
+      {% if kind.resolve == Bool %}
+        state = value ? 1 : 0
+        data = {{name.id.stringify}} == "auto_off_timer" ? Bytes[0x81, state] : state
+      {% elsif kind.resolve == Int32 %}
+        data = value.clamp(0, 100)
+      {% end %}
+      do_send(Command.parse({{name.id.stringify}}), data, **options)
+    end
+  {% end %}
+
+  def do_device_config
+    logger.debug { "Syncronising device state with settings" }
+
+    {% for name, kind in DEVICE_SETTINGS %}
+      %value = setting?({{kind}}, {{name.id.stringify}})
+      {{name.id}}(%value) unless %value.nil?
+    {% end %}
+  end
+
+  enum ResponseStatus
+    Ack = 0x41 # A
+    Nak = 0x4e # N
+  end
+
+  def received(data, task)
+    hex = data.hexstring
+    logger.debug { "Samsung sent: #{hex}" }
+    data = data.map(&.to_i).to_a
+
+    # Calculate checksum of response
+    checksum = data[1..-2].reduce(&.+)
+
+    # Pop also removes the checksum from the response here
+    if data.pop != checksum
+      logger.error { "Invalid checksum\nChecksum should be: #{checksum.to_s(16)}" }
+      return task.try &.retry
+    end
+
+    status = ResponseStatus.from_value(data[4])
+    command = Command.from_value(data[5])
+    values = data[6..-1]
+    value = values.first
+
+    case status
+    when .ack?
+      case command
+      when .status?
+        self[:hard_off]   = values[0] == 0
+        self[:power]      = false if self[:hard_off]
+        self[:volume]     = values[1]
+        self[:audio_mute] = values[2] == 1
+        self[:input]      = Inputs.from_value(values[3]).to_s
```

Comment:
I _think_ that enum values should have `#to_s` invoked automatically. If this is the case, you should be safe to drop the `.to_s` here along with other spots that state is being published.

Reply:
oh yes that appears to be the case. Thanks, that'll be a nice cleanup